src/core/libmaven/PeakDetector.cpp

```diff
@@ -214,52 +216,174 @@ void PeakDetector::processMassSlices(const vector<Compound*>& identificationSet)
+            for (auto scan : sample->scans) {
+                if (scan->mslevel == 2)
+                    scanVector.push_back(scan);
+            }
+            sort(scanVector.begin(), scanVector.end(),
+                 [](const Scan* lhs, const Scan* rhs) {
+                     return lhs->precursorMz < rhs->precursorMz;
+                 });
+            allMs2Scans.insert(make_pair(sample, scanVector));
+        }
+    }
 
-                if (c->precursorMz == 0 || c->productMz == 0) {
+    auto addSlicesWithAdducts = [&](float compoundMass,
+                                    Compound* compound,
+                                    vector<Compound*> compoundVector) {
+        for(auto adduct : adductList) {
+            if (SIGN(adduct->getCharge()) != SIGN(mavenParameters->ionizationMode))
+                continue;
 
-                    //Calculating the mzmin and mzmax
-                    int charge = mavenParameters->getCharge(c);
-                    bool success  = \
-                    slice->calculateMzMinMax(mavenParameters->compoundMassCutoffWindow, charge);
-                    if (!success) continue;
+            mzSlice* slice = new mzSlice();
+            slice->mz = adduct->computeAdductMz(compoundMass);
+            slice->compound = compound; // TODO: this assignment is meaningless.
+            slice->compoundVector = compoundVector;
+            slice->adduct = adduct;
+            bool success =
+                slice->calculateMzMinMax(mavenParameters->compoundMassCutoffWindow,
+                                         adduct->getCharge());
+            if (!success)
+                continue;
+            slice->calculateRTMinMax(mavenParameters->matchRtFlag,
+                                     mavenParameters->compoundRTWindow);
+            slice->setSRMId();
+
+            if (mavenParameters->matchFragmentationFlag) {
+                bool keepSlice = false;
+                for (mzSample* sample : mavenParameters->samples) {
+                    auto it = allMs2Scans.find(sample);
+                    if (it != allMs2Scans.end()) {
+                        auto scanVector = it->second;
+                        auto low = lower_bound(scanVector.begin(),
+                                               scanVector.end(),
+                                               slice->mzmin,
+                                               [](const Scan* scan,
+                                                   const double &val) {
+                                                   return scan->precursorMz < val;
+                                               });
+                        for (auto scanIter = low;
+                             scanIter != scanVector.end();
+                             scanIter++) {
+                            Scan* scan = *scanIter;
+                            if (scan->precursorMz >= slice->mzmin
+                                && scan->precursorMz <= slice->mzmax) {
```

Comment:
Let's not worry about reducing the number of adduct slices at this stage. Adduct groups don't need to have fragmentation information though even if the `fragmentationMatchFlag` is on. As long as the parent ion passes our thresholds, the adduct groups can be created regardless of fragmentation information.

Reply:
Done. I've kept the `filterAdducts` method. I think its still required. But adduct slices are now only created from any detected peak-groups.